packages/node_modules/cerebral/src/utils.js

```diff
@@ -276,12 +276,61 @@ export function createDummyController(state = {}, signals = {}) {
   }
 }
 
+export function copyAndFreezeObject(rootObject) {
+  function makeCopy(value) {
+    if (Array.isArray(value)) {
+      return value.slice()
+    } else if (isObject(value)) {
+      return Object.assign({}, value)
+    }
+
+    return value
+  }
+
+  function freezeObject(object) {
+    if (!Object.isFrozen(object) && isComplexObject(object)) {
+      for (const key in object) {
+        // Properties might not be writable, but then there
+        // is not reason to freeze its value either
+        try {
+          object[key] = freezeObject(object[key])
+        } catch (e) {}
+      }
+
+      return Object.freeze(makeCopy(object))
+    }
+
+    return object
+  }
+
+  return freezeObject(rootObject)
+}
+
+export function addCerebralStateKey(object) {
```

Comment:
This is needed to identify when developers passes objects and arrays from the Cerebral state tree as props down to other Cerebral components. That is not the ideal solution. All objects/arrays should in these scenarios rather be directly connected.

This was already there in previous version, but based on frozen or not. This was rather fragile, as it turns out also children of React is frozen. No matter, freezing is not part of model anymore so now it rather adds a "hidden key" to allow the `View` to identify these passing of Cerebral objects/arrays as props from one connected component to an other.